modules/siddhi-extensions/kf/src/main/java/org/wso2/siddhi/extension/kf/KalmanFilter.java

```diff
@@ -0,0 +1,245 @@
+                if (attributeExpressionExecutors[1].getReturnType() != Attribute.Type.DOUBLE) {
+                    throw new ExecutionPlanValidationException("Invalid parameter type found for the second argument " +
+                                                               "of kf:kalmanFilter() function, required " +
+                                                               Attribute.Type.DOUBLE + ", but found " +
+                                                               attributeExpressionExecutors[1].getReturnType().toString());
+                }
+            }
+            if (attributeExpressionExecutors.length == 4) {
+                if (attributeExpressionExecutors[2].getReturnType() != Attribute.Type.DOUBLE) {
+                    throw new ExecutionPlanValidationException("Invalid parameter type found for the third argument " +
+                                                               "of kf:kalmanFilter() function, required " +
+                                                               Attribute.Type.DOUBLE + ", but found " +
+                                                               attributeExpressionExecutors[1].getReturnType().toString());
+                }
+                if (attributeExpressionExecutors[3].getReturnType() != Attribute.Type.LONG) {
+                    throw new ExecutionPlanValidationException("Invalid parameter type found for the fourth argument " +
+                                                               "of kf:kalmanFilter() function, required " +
+                                                               Attribute.Type.LONG + ", but found " +
+                                                               attributeExpressionExecutors[1].getReturnType().toString());
+                }
+            }
+        }
+    }
+
+    @Override
+    protected Object execute(Object[] data) {
+        if (data[0] == null) {
+            throw new ExecutionPlanRuntimeException("Invalid input given to kf:kalmanFilter() " +
+                                                    "function. First argument should be a double");
+        }
+
+        if (data.length == 1) {
+            double measuredValue = (Double) data[0]; //to remain as the initial state
+            if (transition == 0) {
+                transition = 1;
+                variance = 1000;
+                measurementNoiseSD = 0.001d;
+                prevEstimatedValue = measuredValue;
+            }
+            prevEstimatedValue = transition * prevEstimatedValue;
+            double kalmanGain = variance / (variance + measurementNoiseSD);
+            prevEstimatedValue = prevEstimatedValue + kalmanGain * (measuredValue - prevEstimatedValue);
+            variance = (1 - kalmanGain) * variance;
+            return prevEstimatedValue;
+        } else if (data.length == 2) {
+            if (data[1] == null) {
```

Comment:
you have to also need to check the return type of the value is double. if its not double then you have to send an error to that too